### PR TITLE
docs(react-infobutton): Adding component's spec

### LINKED ISSUE #25061: InfoButton: Write and merge spec
As InfoButton is started, first step is to write a spec and merge it.

Related #25062

### PR DESCRIPTION
**Pull request by sopranopillow**
This PR adds the spec for InfoButton. Note: this is still a WIP and will be updated after a design review of the component.

[Preview](https://github.com/microsoft/fluentui/blob/9520d7c904fdd8709ab1e3aca75ae7eefc4d2904/packages/react-components/react-infobutton/Spec.md)

## Related Issue(s)

Fixes #25061

### CODE REVIEW

**Comment by fabricteam** [comment]
## 📊 Bundle size report

<details>
<summary>Unchanged fixtures</summary>

| Package & Exports                                                                                      |       Size (minified/GZIP) |
| ------------------------------------------------------------------------------------------------------ | -------------------------: |
| <samp>react-infobutton</samp> <br /> <abbr title='bundle-size/InfoButton.fixture.js'>InfoButton</abbr> | `7.742 kB`<br />`3.265 kB` |

</details>
<sub>🤖 This report was generated against <a href='https://github.com/microsoft/fluentui/commit/fec73cb3ed23fb29387af6a9641f1012b129db52'>fec73cb3ed23fb29387af6a9641f1012b129db52</a></sub>
<div id="bundleSizeComment9423"/>

File: packages/react-components/react-infobutton/Spec.md [hunk 1]
```diff
@@ -1,52 +1,138 @@
...
+import type { PopoverProps, PopoverSurface } from '@fluentui/react-popover';
+import type { ComponentProps, ComponentState, Slot } from '@fluentui/react-utilities';
+
+export type InfoButtonSlots = {
+  root: NonNullable<Slot<PopoverProps>>;
+
+  /**
+   * The button that triggers the Popover.
+   */
+  trigger: NonNullable<Slot<typeof Button>>;
```
Comment (layershifter) [inline packages/react-components/react-infobutton/Spec.md:58]: Let's avoid calling this `trigger` to avoid confusion with the triggers pattern.

File: packages/react-components/react-infobutton/Spec.md [hunk 2]
```diff
@@ -1,52 +1,138 @@
...
 
-- _**Public**_
-- _**Internal**_
-- _**DOM** - how the component will be rendered as HTML elements_
+_**Public**_
 
-## Migration
+```jsx
+<InfoButton content="This is some additional information." />
+```
```
Comment (layershifter) [inline packages/react-components/react-infobutton/Spec.md:85]: Is it possible to attach a `Tooltip` to it? I.e.

```jsx
<Tootlip>
  <InfoButton content="This is some additional information." />
</Tooltip>
```

File: packages/react-components/react-infobutton/Spec.md [hunk 3]
```diff
@@ -1,52 +1,138 @@
...
+  /**
+   * The content to be displayed in the Popover.
+   */
+  content: NonNullable<Slot<typeof PopoverSurface>>;
+};
+
+/**
+ * InfoButton Props
+ */
+export type InfoButtonProps = Omit<ComponentProps<Partial<InfoButtonSlots>>, 'children'>;
```
Comment (layershifter) [inline packages/react-components/react-infobutton/Spec.md:69]: Is there is no `children`, how can I customize `button`'s content? `<InfoButton trigger={{ children: 'X' }} />`?

File: packages/react-components/react-infobutton/Spec.md [hunk 4]
```diff
@@ -1,52 +1,138 @@
...
+
+/**
+ * State used in rendering InfoButton
+ */
+export type InfoButtonState = ComponentState<InfoButtonSlots> & {
+  popoverOpen: boolean;
```
Comment (layershifter) [inline packages/react-components/react-infobutton/Spec.md:75]: Will be this exposed in `props`? If no, how can I control `Popover`'s state?

File: packages/react-components/react-infobutton/Spec.md [hunk 5]
```diff
@@ -1,52 +1,138 @@
 # @fluentui/react-infobutton Spec
 
+Convergence epic issue: [#25062](https://github.com/microsoft/fluentui/issues/25062)
+
 ## Background
```
Comment (layershifter) [inline packages/react-components/react-infobutton/Spec.md:5]: Can you please add some visuals to show how component looks?

File: packages/react-components/react-infobutton/Spec.md [hunk 6]
```diff
@@ -1,52 +1,138 @@
...
+Because the Popover isn't always visible, it should not contain information that people must know in order to complete the field.
 
 ## Prior Art
 
 _Include background research done for this component_
 
-- _Link to Open UI research_
-- _Link to comparison of v7 and v0_
-- _Link to GitHub epic issue for the converged component_
+#### Open UI
```
Comment (layershifter) [inline packages/react-components/react-infobutton/Spec.md:15]: This section is not related to OpenUI, it should be under "Prior Art", correct?

File: packages/react-components/react-infobutton/Spec.md [hunk 7]
```diff
@@ -1,52 +1,138 @@
 # @fluentui/react-infobutton Spec
 
+Convergence epic issue: [#25062](https://github.com/microsoft/fluentui/issues/25062)
+
 ## Background
 
-_Description and use cases of this component_
+An InfoButton provides a way for users to get more information about a particular UI element. It is a button that displays an icon and when clicked a Popover displays the additional information.
```
Comment (layershifter) [inline packages/react-components/react-infobutton/Spec.md:7]: Do I correctly understand that this component is basically following?

```jsx
<Popover>
  <PopoverTrigger>
    <Button icon='X' />
  </PopoverTrigger>
  <PopoverSurface>Some content</PopoverSurface>
</Popover>
```

File: packages/react-components/react-infobutton/Spec.md (hunk 2 above)
Comment (sopranopillow) [reply packages/react-components/react-infobutton/Spec.md:85]: I'm not sure why someone would do that in this case, but it should be possible.

File: packages/react-components/react-infobutton/Spec.md (hunk 3 above)
Comment (sopranopillow) [reply packages/react-components/react-infobutton/Spec.md:69]: Correct, it would be through the slot. This omit is only removing the children for the root slot (Popover). Another option is to make children be the children of content, but I decided against that due to content being a slot itself.

File: packages/react-components/react-infobutton/Spec.md (hunk 4 above)
Comment (sopranopillow) [reply packages/react-components/react-infobutton/Spec.md:75]: That can be done directly on the props like this:
`<InfoButton open={controlledOpen} />`

File: packages/react-components/react-infobutton/Spec.md (hunk 7 above)
Comment (sopranopillow) [reply packages/react-components/react-infobutton/Spec.md:7]: After an update, we are not using the v9 button, but the native button. Other than that, you are correct. It's a pretty simple component, but will bring an opinionated approach to it required for Field.

File: packages/react-components/react-infobutton/Spec.md [hunk 8]
```diff
@@ -1,63 +1,143 @@
...
+
+<!-- on document.body -->
+<div role="tooltip" class="fui-PopoverSurface fui-InfoButton__content">This is some additional information.</div>
+```
+
+## Migration
 
-- _Migration from v8_
-- _Migration from v0_
+There's no migration guide as `v0` and `v8` do not have an InfoButton component.
```
Comment (kolaps33) [inline packages/react-components/react-infobutton/Spec.md:110]: Hello :)
there was no similar component in `v0` or in 'v8' but one version exists in the TMP repository(in MS Teams code)

Implementation accessibility details are following:
- icon is focusable
- icon has role=img
- with tooltip is icon connected via "aria-describedby" approach
- icon has label "More information"
- tooltip is displayed when icon receive focus
https://domoreexp.visualstudio.com/Teamspace/_git/teams-modular-packages/pullrequest/477885

This approach was approved by accessibility PM for MS Teams(Brett Humphrey).

I believe we should be aligned together with v9 and MS Teams component.

**Comment by layershifter** [comment]
@sopranopillow I was thinking about this spec and here is a proposal from my side...

## Make `InfoButton` a button

In the spec currently `button` is a slot at `InfoButton` component. It looks a bit strange as even our form components have a concept our primary slot and behave differently. The only components that are custom are hosts (`Menu`, `Popover`, `Tooltip`, etc.) and triggers (`MenuTrigger`, etc.). However, both hosts and triggers don't render DOM elements.


So, what is a proposed change? In terms of render template:

```tsx
  <Popover>
    <PopoverTrigger>
      <slots.root {...slotProps.root} />
    </PopoverTrigger>
    <PopoverContent />
  </Popover>
```

Instead of `Popover` being a root - the proposal is to have a `button` as a root (this solves this concern https://github.com/microsoft/fluentui/pull/25251#discussion_r996742657).

This change allows following:
- Use `Tooltip`s and other components:
  ```tsx
    <Tooltip content="A tooltip">
       <InfoButton>
         A button
       </InfoButton>
     </Tooltip>
  ```
- Use props directly (I mean `className`, `style`, `ref` and etc.)
   ```tsx
     <InfoButton className='foo' ref={buttonRef} />
     // over
    <InfoButton className='foo' ref={buttonRef} /> // 💥 does not work
    <InfoButton button={{ className: 'foo', ref: buttonRef }} /> // 😕 works, but not consistent with other components
   ```
- We can pass `children` to a component that makes usages intuitive
   ```tsx
   <InfoButton>Content goes here</InfoButton>
   ```

### What to do with `Popover` then?

It's a good question for that proposal 👆

We don't need to configure `PopoverTrigger`, but we need to provide an access to `Popover` and `PopoverSurface`. I was thinking about `popover` & `popoverSurface` slots:

```tsx
<InfoButton popover={{ open, positioning }} popoverSurface={{ children: 'Content' }}>
   A button
</InfoButton>
```

- where `popover` is an object with slots and cannot have shorthand form and `children`
- where `popoverSurface` is a regular slot

Another option is combine that with a current proposal and keep `Popover` props directly on `InfoButton`:

```tsx
<InfoButton open, positioning={{ /* ... */ }} popoverSurface={{ children: 'Content' }}>
   A button
</InfoButton>
```

---

From my POV it fits more into v9 API design, WDYT?

**Comment by sopranopillow** [comment]
> @sopranopillow I was thinking about this spec and here is a proposal from my side...
> 
> ## Make `InfoButton` a button
> In the spec currently `button` is a slot at `InfoButton` component. It looks a bit strange as even our form components have a concept our primary slot and behave differently. The only components that are custom are hosts (`Menu`, `Popover`, `Tooltip`, etc.) and triggers (`MenuTrigger`, etc.). However, both hosts and triggers don't render DOM elements.
> 
> So, what is a proposed change? In terms of render template:
> 
> ```tsx
>   <Popover>
>     <PopoverTrigger>
>       <slots.root {...slotProps.root} />
>     </PopoverTrigger>
>     <PopoverContent />
>   </Popover>
> ```
> 
> Instead of `Popover` being a root - the proposal is to have a `button` as a root (this solves this concern [#25251 (comment)](https://github.com/microsoft/fluentui/pull/25251#discussion_r996742657)).
> 
> This change allows following:
> 
> * Use `Tooltip`s and other components:
>   ```tsx
>     <Tooltip content="A tooltip">
>        <InfoButton>
>          A button
>        </InfoButton>
>      </Tooltip>
>   ```
> * Use props directly (I mean `className`, `style`, `ref` and etc.)
>   ```tsx
>     <InfoButton className='foo' ref={buttonRef} />
>     // over
>    <InfoButton className='foo' ref={buttonRef} /> // 💥 does not work
>    <InfoButton button={{ className: 'foo', ref: buttonRef }} /> // 😕 works, but not consistent with other components
>   ```
> * We can pass `children` to a component that makes usages intuitive
>   ```tsx
>   <InfoButton>Content goes here</InfoButton>
>   ```
> 
> ### What to do with `Popover` then?
> It's a good question for that proposal 👆
> 
> We don't need to configure `PopoverTrigger`, but we need to provide an access to `Popover` and `PopoverSurface`. I was thinking about `popover` & `popoverSurface` slots:
> 
> ```tsx
> <InfoButton popover={{ open, positioning }} popoverSurface={{ children: 'Content' }}>
>    A button
> </InfoButton>
> ```
> 
> * where `popover` is an object with slots and cannot have shorthand form and `children`
> * where `popoverSurface` is a regular slot
> 
> Another option is combine that with a current proposal and keep `Popover` props directly on `InfoButton`:
> 
> ```tsx
> <InfoButton open, positioning={{ /* ... */ }} popoverSurface={{ children: 'Content' }}>
>    A button
> </InfoButton>
> ```
> 
> From my POV it fits more into v9 API design, WDYT?

@layershifter Love it! I think this is a great candidate for a recipe on how to create a PopupButton (name not set) the Fluent v9 way!

File: packages/react-components/react-infobutton/Spec.md [hunk 9]
```diff
@@ -1,63 +1,143 @@
...
-- _Link to comparison of v7 and v0_
-- _Link to GitHub epic issue for the converged component_
+- v0 does not have an InfoButton component.
+- v8 does not have an InfoButton component, but does have a TextField component that has an example of how to add an info button to a TextField, see Open UI section above for more details.
 
 ## Sample Code
 
-_Provide some representative example code that uses the proposed API for the component_
+```jsx
+<InfoButton content="This is some additional information." />
```
Comment (spmonahan) [inline packages/react-components/react-infobutton/Spec.md:33]: One of the images shows a link in the popover. Would be nice to have a code sample showing you can use JSX in the `content` prop to make it clear we `InfoButton` supports that.

File: packages/react-components/react-infobutton/Spec.md [hunk 10]
```diff
@@ -1,63 +1,143 @@
...
 ## Background
 
-_Description and use cases of this component_
+An InfoButton provides a way for users to get more information about a particular UI element. It is a button that displays an icon and when clicked a Popover displays the additional information.
+
+Because the Popover isn't always visible, it should not contain information that people must know in order to complete the field.
+
+### Anatomy
+
+![Anatomy](./etc/images/anatomy.png)
```
Comment (spmonahan) [inline packages/react-components/react-infobutton/Spec.md:13]: This image shows the popover in the "above-start" position. Will `InfoButton` support positioning of the popover?

Comment (sopranopillow) [reply packages/react-components/react-infobutton/Spec.md:13]: Yes, that will be done through the root as it takes the popover props as well. The updated types now have the Popover props.

File: packages/react-components/react-infobutton/Spec.md [hunk 11]
```diff
@@ -1,63 +1,141 @@
...
+#### Props
+
+```ts
+export type InfoButtonSlots = {
+  root: NonNullable<Slot<'button'>>;
+
+  /**
+   * The PopoverSurface to be displayed when the button is pressed.
+   */
+  popoverSurface: NonNullable<Slot<typeof PopoverSurface>>;
```
Comment (behowell) [inline packages/react-components/react-infobutton/Spec.md:55]: Can we call this `content` so it is similar to Tooltip?

File: packages/react-components/react-infobutton/Spec.md [hunk 12]
```diff
@@ -1,63 +1,141 @@
...
+};
+
+/**
+ * InfoButton Props
+ */
+export type InfoButtonProps = ComponentProps<Partial<InfoButtonSlots>> & Omit<PopoverProps, 'children'>;
```
Comment (behowell) [inline packages/react-components/react-infobutton/Spec.md:61]: This should pick specific popover props, rather than taking all of them (minus children). The code in useInfoButton needs to know about each Popover prop that it's forwarding. If a new prop is added to Popover in the future, that would implicitly add it to InfoButton, but it wouldn't be functional. So it'd be safer to explicitly list the Popover props that are supported by InfoButton.

Comment (behowell) [reply packages/react-components/react-infobutton/Spec.md:61]: That said, I don't think we typically do prop forwarding like this. It'd be a little less ergonomic, but I think having a `popover` slot would be more in line with the v9 API design. That would also avoid having to manually pick and forward props to the popover.

File: packages/react-components/react-infobutton/Spec.md [hunk 13]
```diff
@@ -1,63 +1,141 @@
...
+    </>
+  }
+/>
+```
+
+_**Internal**_
+
+```jsx
+return (
+  <Popover {...(state as Omit<PopoverProps, 'children'>)}>
```
Comment (behowell) [inline packages/react-components/react-infobutton/Spec.md:83]: I suppose this is more of a code review comment, but we definitely don't want to blindly spread state like this. You'd need to write out each of the props being forwarded. (Or make `popover` a slot like in my other comment)

File: packages/react-components/react-infobutton/Spec.md [hunk 14]
```diff
@@ -1,63 +1,141 @@
...
-- Decide whether to use **native element** or folow **ARIA** and provide reasons
-- Identify the **[ARIA](https://www.w3.org/TR/wai-aria-practices-1.2/) pattern** and, if the component is listed there, follow its specification as possible.
-- Identify accessibility **variants**, the `role` ([ARIA roles](https://www.w3.org/TR/wai-aria-1.1/#role_definitions)) of the component, its `slots` and `aria-*` props.
-- Describe the **keyboard navigation**: Tab Oder and Arrow Key Navigation. Describe any other keyboard **shortcuts** used
-- Specify texts for **state change announcements** - [ARIA live regions
-  ](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Live_Regions) (number of available items in dropdown, error messages, confirmations, ...)
-- Identify UI parts that appear on **hover or focus** and specify keyboard and screen reader interaction with them
-- List cases when **focus** needs to be **trapped** in sections of the UI (for dialogs and popups or for hierarchical navigation)
-- List cases when **focus** needs to be **moved programatically** (if parts of the UI are appearing/disappearing or other cases)
+- `role="tooltip"` is used on the PopoverSurface.
```
Comment (behowell) [inline packages/react-components/react-infobutton/Spec.md:134]: Do we then also need to use `aria-describedby` or `aria-labeledby` on the button itself?

Comment (sopranopillow) [reply packages/react-components/react-infobutton/Spec.md:134]: This isn't set yet, once we iron it out, I'll update the spec.

File: packages/react-components/react-infobutton/Spec.md [hunk 15]
```diff
@@ -1,63 +1,151 @@
...
-- _**Public**_
-- _**Internal**_
-- _**DOM** - how the component will be rendered as HTML elements_
+_**Public**_
+
+```jsx
+<InfoButton
+  content={
+    <>
+      Popover above-start lorem ipsum dolor sit amet consectetur.{' '}
```
Comment (micahgodbolt) [inline packages/react-components/react-infobutton/Spec.md:82]:
```suggestion
      Popover above-start lorem ipsum dolor sit amet consectetur.
```

File: packages/react-components/react-infobutton/Spec.md [hunk 16]
```diff
@@ -1,63 +1,151 @@
...
+      <Link href="https://react.fluentui.dev">Learn more</Link>
```
Comment (micahgodbolt) [inline packages/react-components/react-infobutton/Spec.md:83]: Is focus going to be handled to allow actionable items in the popover? I don't recall that being addressed in the description above.

Comment (micahgodbolt) [reply packages/react-components/react-infobutton/Spec.md:83]: I see it mentioned below, but it might be good to call out in the initial description above (can contain focusable content)

File: packages/react-components/react-infobutton/Spec.md (hunk 7 above)
Comment (micahgodbolt) [reply packages/react-components/react-infobutton/Spec.md:7]: This paragraph should be updated to make it obvious that InfoButton can contain focusable content as well (Link/button etc)

File: packages/react-components/react-infobutton/Spec.md [hunk 17]
```diff
@@ -1,63 +1,151 @@
...
-- Specify texts for **state change announcements** - [ARIA live regions
-  ](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Live_Regions) (number of available items in dropdown, error messages, confirmations, ...)
-- Identify UI parts that appear on **hover or focus** and specify keyboard and screen reader interaction with them
-- List cases when **focus** needs to be **trapped** in sections of the UI (for dialogs and popups or for hierarchical navigation)
-- List cases when **focus** needs to be **moved programatically** (if parts of the UI are appearing/disappearing or other cases)
+- `role="tooltip"` is used on the PopoverSurface.
+- Tab order
+  - When tabbing through the page, the button should be a tab stop.
+  - When focused on the button, pressing `Enter` or `Space` should open the Popover.
+    - Focus is not trapped within the Popover, it should stay in the button.
```
Comment (micahgodbolt) [inline packages/react-components/react-infobutton/Spec.md:148]:
```suggestion
    - If the popover does not contain any focusable items, focus should stay on the button.
```

File: packages/react-components/react-infobutton/Spec.md [hunk 18]
```diff
@@ -1,63 +1,151 @@
 # @fluentui/react-infobutton Spec
 
+Convergence epic issue: [#25062](https://github.com/microsoft/fluentui/issues/25062)
+
 ## Background
 
-_Description and use cases of this component_
+An InfoButton provides a way for users to get more information about a particular UI element. It is a button that displays an icon and when clicked a Popover displays the additional information. InfoButton may contain focusable items inside the Popover.
```
Comment (khmakoto) [inline packages/react-components/react-infobutton/Spec.md:7]: _nit:_ wording

```suggestion
An InfoButton provides a way for users to get more information about a particular UI element. It is a button containing an icon that, when clicked, displays a Popover with the additional information. InfoButton may contain focusable items inside the Popover.
```

File: packages/react-components/react-infobutton/Spec.md [hunk 19]
```diff
@@ -1,63 +1,151 @@
...
+| TextField with example on how to add an info button | [v8](https://developer.microsoft.com/en-us/fluentui#/controls/web/textfield)          | This is just an example and on how to implement it.                                                      |
+| ContextualHelp                                      | [React Spectrum](https://react-spectrum.adobe.com/react-spectrum/ContextualHelp.html) | It's a full component and uses a Popover instead of tooltip.                                             |
 
-- _Link to Open UI research_
-- _Link to comparison of v7 and v0_
-- _Link to GitHub epic issue for the converged component_
+#### Comparison with v0 and v8
+
+- v0 does not have an InfoButton component.
+- v8 does not have an InfoButton component, but does have a TextField component that has an example of how to add an info button to a TextField, see Open UI section above for more details.
```
Comment (khmakoto) [inline packages/react-components/react-infobutton/Spec.md:28]:
```suggestion
- v8 does not have an InfoButton component, but does have a TextField component that has an example of how to add an info button to a TextField, see "Prior Art" section above for more details.
```

File: packages/react-components/react-infobutton/Spec.md [hunk 20]
```diff
@@ -1,63 +1,151 @@
...
-_Describe what will need to be done to upgrade from the existing implementations:_
-
-- _Migration from v8_
-- _Migration from v0_
+There's no migration guide as `v0` and `v8` do not have an InfoButton component.
 
 ## Behaviors
 
-_Explain how the component will behave in use, including:_
+> Note: The behavior will change when used in conjunction with the `Field` component. It is mentioned that when a form has many infobuttons, it may be better to not make a full tab stop in the button. The idea is to have each field have a "shortcut" that will let you focus on the infobutton. This will be implemented by Field if needed.
```
Comment (khmakoto) [inline packages/react-components/react-infobutton/Spec.md:122]:
```suggestion
> Note: The behavior will change when used in conjunction with the `Field` component. It is mentioned that when a form has many InfoButtons, it may be better to not make a full tab stop in the button. The idea is to have each field have a "shortcut" that will let you focus on the infobutton. This will be implemented by Field if needed.
```

**Review by layershifter — APPROVED** [review]

File: packages/react-components/react-infobutton/Spec.md [hunk 21]
```diff
@@ -1,63 +1,151 @@
...
+```
 
 ## Variants
 
-_Describe visual or functional variants of this control, if applicable. For example, a slider could have a 2D variant._
+Visual (only the button has a variant)
+
+- `transparent` (default)
+
+![transparent](./etc/images/transparent-infobutton.png)
```
Comment (behowell) [inline packages/react-components/react-infobutton/Spec.md:42]: InfoButtons are _always_ transparent right? So it's not technically a variant 🙂 it's just how InfoButtons look.

**Review by behowell — APPROVED** [review]

File: packages/react-components/react-infobutton/Spec.md (hunk 21 above)
Comment (sopranopillow) [reply packages/react-components/react-infobutton/Spec.md:42]: It hasn't been finalized, but we might add subtle. I'll remove it for now!
